lib/pbio/drv/usb/usb_ev3.c

```diff
@@ -875,41 +881,97 @@ static void usb_device_intr(void) {
-
-            // TODO: Remove this echo test
-            unsigned int i;
-            for (i = 0; i < usb_rx_sz; i++) {
-                ep1_tx_response_buf[i] = ep1_rx_buf[i] + 1;
-            }
-            for (; i < 512; i++) {
-                ep1_tx_response_buf[i] = 0xaa;
+            pbio_pybricks_error_t result;
+            bool usb_send_response = false;
+            // Skip empty commands, or commands sent when previous response is still busy
+            if (usb_rx_sz && !usb_tx_response_is_not_ready) {
+                switch (ep1_rx_buf[0]) {
+                    case PBIO_PYBRICKS_OUT_EP_MSG_SUBSCRIBE:
+                        pbdrv_usb_is_events_subscribed = ep1_rx_buf[1];
+                        ep1_tx_response_buf[0] = PBIO_PYBRICKS_IN_EP_MSG_RESPONSE;
+                        pbio_set_uint32_le(&ep1_tx_response_buf[1], PBIO_PYBRICKS_ERROR_OK);
+                        usb_send_response = true;
+                        break;
+                    case PBIO_PYBRICKS_OUT_EP_MSG_COMMAND:
+                        result = pbsys_command(ep1_rx_buf + 1, usb_rx_sz - 1);
+                        ep1_tx_response_buf[0] = PBIO_PYBRICKS_IN_EP_MSG_RESPONSE;
+                        pbio_set_uint32_le(&ep1_tx_response_buf[1], result);
+                        usb_send_response = true;
+                        break;
+                }
             }
 
-            (void)ep1_tx_status_buf;
-            (void)ep1_tx_stdout_buf;
-
-            unsigned int tx_sz = pbdrv_usb_is_usb_hs ? PYBRICKS_EP_PKT_SZ_HS : PYBRICKS_EP_PKT_SZ_FS;
-            if (!usb_tx_response_is_not_ready) {
+            if (usb_send_response) {
                 usb_tx_response_is_not_ready = true;
-                usb_setup_tx_dma_desc(CPPI_DESC_TX_RESPONSE, ep1_tx_response_buf, tx_sz);
+                usb_setup_tx_dma_desc(CPPI_DESC_TX_RESPONSE, ep1_tx_response_buf, USB_RESPONSE_ACTUAL_SZ);
             }
 
             // Re-queue RX buffer after processing is complete
             usb_rx_is_ready = false;
             usb_setup_rx_dma_desc();
         }
+
+        // Send status flags if they've changed (and we can)
+        new_status_flags = pbsys_status_get_flags();
+        if (pbdrv_usb_is_events_subscribed && !usb_tx_status_is_not_ready && new_status_flags != prev_status_flags) {
+            ep1_tx_status_buf[0] = PBIO_PYBRICKS_IN_EP_MSG_EVENT;
+            uint32_t usb_status_sz = 1 + pbsys_status_get_status_report(&ep1_tx_status_buf[1]);
+
+            usb_tx_status_is_not_ready = true;
+            usb_setup_tx_dma_desc(CPPI_DESC_TX_STATUS, ep1_tx_status_buf, usb_status_sz);
+
+            prev_status_flags = new_status_flags;
+        }
+
+        // Handle timeouts
+        is_transmitting = usb_tx_response_is_not_ready || usb_tx_status_is_not_ready || usb_tx_stdout_is_not_ready;
+        if (was_transmitting && is_transmitting) {
+            if (pbio_os_timer_is_expired(&tx_timeout_timer)) {
```

Comment:
Are we checking for timeout before setting the timer?

Reply:
No, the timer gets set in the `else if` later, which this code has to pass through before both flags can become true.